driver/src/main/scala/api/Driver.scala

```diff
@@ -0,0 +1,239 @@
+    SystemControl(system) match {
+      case Success(TimedSystemControl(close)) => { timeout =>
+        close(timeout) match {
+          case Failure(cause) => Future.failed[Unit](cause)
+          case _              => Future.successful({})
+        }
+      }
+
+      case Success(AsyncSystemControl(close)) =>
+        { _ => close() }
+
+      case Failure(cause) => { _ => Future.failed[Unit](cause) }
+    }
+
+  protected final val supervisorName = s"Supervisor-${Driver.nextCounter}"
+  private[reactivemongo] final val supervisorActor =
+    system.actorOf(Props(new SupervisorActor(this)), supervisorName)
+
+  protected final val connectionMonitors =
+    MutableMap.empty[ActorRef, MongoConnection]
+
+  /**
+   * Closes this driver (and all its connections and resources).
+   * Will wait until the timeout for proper closing of connections before forcing hard shutdown.
+   */
+  protected final def askClose(timeout: FiniteDuration)(implicit executionContext: ExecutionContext): Future[Unit] = {
+    logger.info(s"[$supervisorName] Closing instance of ReactiveMongo driver")
+    // Tell the supervisor to close.
+    // It will shut down all the connections and monitors
+    (supervisorActor ? Close)(Timeout(timeout)).recover {
+      case err =>
+        logger.warn(s"[$supervisorName] Failed to close connections within timeout. Continuing closing of ReactiveMongo driver anyway.", err)
+    } // and then shut down the ActorSystem as it is exiting.
+      .flatMap(_ => systemClose(Some(timeout)))
+  }
+
+  /**
+   * Creates a new MongoConnection.
+   *
+   * @param nodes $nodesParam
+   * @param authentications $authParam
+   * @param name $connectionNameParam
+   * @param options $optionsParam
+   */
+  protected final def askConnection(
+    nodes: Seq[String],
+    options: MongoConnectionOptions,
+    authentications: Seq[Authenticate],
+    name: Option[String]): Future[MongoConnection] = {
+
+    val nm = name.getOrElse(s"Connection-${+Driver.nextCounter}")
+
+    // TODO: Passing ref to MongoDBSystem.history to AddConnection
+    lazy val dbsystem: MongoDBSystem = options.authMode match {
+      case CrAuthentication => new LegacyDBSystem(
+        supervisorName, nm, nodes, authentications, options)
+
+      case _ => new StandardDBSystem(
+        supervisorName, nm, nodes, authentications, options)
+    }
+
+    val mongosystem = system.actorOf(Props(dbsystem), nm)
+
+    def connection = (supervisorActor ? AddConnection(
+      nm, nodes, options, mongosystem))(Timeout(10, SECONDS))
+
+    logger.info(s"[$supervisorName] Creating connection: $nm")
+
+    import system.dispatcher
+
+    connection.mapTo[MongoConnection].map { c =>
+      // TODO: Review
+      c.history = () => dbsystem.internalState
+      c
+    }
+  }
+
+  private case class AddConnection(
+    name: String,
+    nodes: Seq[String],
+    options: MongoConnectionOptions,
+    mongosystem: ActorRef)
+  
+  // For testing only
+  private[api] def addConnectionMsg(
```

Comment:
@jbgi Need to update [`src/test/scala/ApiTest.scala`](https://github.com/jbgi/ReactiveMongo/blob/d3bd78caa90a8d6651715e4b09bad02ab5120bc0/driver/src/test/scala/ApiTesting.scala#L35) using that